File: Java/LinkedList/01_Singly_Linear_LinkedList_Java/SinglyLL_008.java
/*
===============================================================================

Program Name : Singly Linear Linked List - Linked List Class Initialization
Program Number  : 008
Description     : This program demonstrates how to create a
                  separate SinglyLL class to represent a
                  Singly Linear Linked List in Java.

                  The constructor initializes the first
                  reference to null and the node count
                  (iCount) to zero, preparing an empty
                  linked list for future operations.
Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================
Concepts Used:

1. Classes and Objects
2. Self-Referential Class
3. Constructors
4. Encapsulation
5. Object-Oriented Programming
6. Instance Variables
7. Object References
8. this Keyword

Key Learning Points:

- The SinglyLL class represents the linked list.
- first stores the reference to the first node.
- iCount stores the total number of nodes.
- Constructors initialize linked list objects automatically.
- Encapsulation combines data and operations into a single class.
- Every new linked list object starts in an empty state.

===============================================================================
*/
class node
{
    public int data;           // Data value stored in node
    public node next;          // Reference to next node

    // Parameterized Constructor
    node(int no)
    {
        this.data = no;        // Initialize data with parameter
        this.next = null;      // Initialize next to null
    }
}

class SinglyLL
{
    public node first;         // Reference to first node (head)
    public int iCount;         // Counter for number of nodes

    // Constructor for SinglyLL class
    public SinglyLL()
    {
        System.out.println("Inside Constructor");
        this.first = null;     // No nodes initially
        this.iCount = 0;       // Counter starts at 0
    }
}

class SinglyLL008
{
    public static void main(String A[])
    {
        // Create SinglyLL object - constructor runs automatically
        SinglyLL sobj = new SinglyLL();
    }
}

/*
================================================================================
Expected Output:
================================================================================
Inside Constructor

================================================================================
Memory Diagram:
================================================================================

Stack:
┌──────────────────────────┐
│ sobj (reference variable)│
│ Points to SinglyLL object│
└──────────────────────────┘
         │
         ▼
Heap:
┌─────────────────────────┐
│ SinglyLL Object:        │
│ first: null             │
│ iCount: 0               │
└─────────────────────────┘

Constructor Execution:
new SinglyLL() → Allocate memory on heap
                → Constructor SinglyLL() runs
                → Print "Inside Constructor"
                → this.first = null
                → this.iCount = 0
                → Return reference to sobj

================================================================================
Pointer/Logic Explanation:
================================================================================

Linked List Class

SinglyLL

Instance Variables

• first
  - Stores the reference of the first node.
  - Initially null.

• iCount
  - Stores the total number of nodes.
  - Initially 0.

Constructor

public SinglyLL()

• Executes automatically when a SinglyLL object is created.
• Initializes first to null.
• Initializes iCount to 0.

===============================================================================
Operations Performed
===============================================================================
1. Defined a self-referential node class.
2. Created a parameterized constructor for node initialization.
3. Created the SinglyLL class.
4. Declared first as the head reference.
5. Declared iCount to store the node count.
6. Initialized first to null.
7. Initialized iCount to 0.
8. Created a SinglyLL object.
9. Executed the constructor automatically.

================================================================================
Time Complexity:
================================================================================
Constructor Execution  : O(1) - Simple initialization
Object Creation        : O(1) - Allocate memory
Variable Access        : O(1) - Direct field access
Space Complexity       : O(1) - Single SinglyLL object

================================================================================
Conclusion:
================================================================================

This program demonstrates how a Singly Linear Linked List
can be represented using a separate class in Java. The
constructor initializes the linked list by setting the
first reference to null and the node count to zero.

This object-oriented design forms the foundation for
implementing linked list operations such as insertion,
deletion, traversal, and searching in the upcoming programs.

================================================================================
*/

File: C++/LinkedList/01_Singly_Linear_LinkedList_CPP/SinglyLL_008.cpp
/*
===============================================================================
Program Name    : Singly Linked List using Scope Resolution Operator
Program Number  : 008 
Description     : This program demonstrates how member functions of a Singly
                  Linked List class can be declared inside the class and
                  defined outside using the Scope Resolution Operator (::).
                  This improves code organization, readability, and follows
                  standard Object-Oriented Programming practices.
Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Object-Oriented Programming (OOP)
2. Class and Object
3. Constructor
4. Scope Resolution Operator (::)
5. Function Declaration
6. Function Definition
7. Encapsulation
8. Data Hiding

Key Learning Points:
- Member functions can be declared inside the class.
- Function definitions can be written outside the class.
- Scope Resolution Operator (::) connects declarations with definitions.
- Separating interface and implementation improves code readability.
- This approach is commonly used in large C++ applications.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

class SinglyLL
{
    private:

        PNODE first;
        int iCount;

    public:

        SinglyLL();

        void Display();
        int Count();

        void InsertFirst(int iNo);
        void InsertLast(int iNo);
        void InsertAtPos(int iNo,int iPos);

        void DeleteFirst();
        void DeleteLast();
        void DeleteAtPos(int iPos);
};

/*
===============================================================================
Constructor Definition
===============================================================================
*/

SinglyLL::SinglyLL()
{
    cout<<"Inside Constructor\n";

    this->first = NULL;
    this->iCount = 0;
}

/*
===============================================================================
Member Function Definitions
===============================================================================
*/

void SinglyLL::Display()
{
}

int SinglyLL::Count()
{
    return this->iCount;
}

void SinglyLL::InsertFirst(int iNo)
{
}

void SinglyLL::InsertLast(int iNo)
{
}

void SinglyLL::InsertAtPos(int iNo,int iPos)
{
}

void SinglyLL::DeleteFirst()
{
}

void SinglyLL::DeleteLast()
{
}

void SinglyLL::DeleteAtPos(int iPos)
{
}

int main()
{
    SinglyLL sobj;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Inside Constructor

===============================================================================
Class Representation:
===============================================================================

                SinglyLL
      +-----------------------------+
      | first   (private)           |
      | iCount  (private)           |
      +-----------------------------+
      | Display()                   |
      | Count()                     |
      | InsertFirst()               |
      | InsertLast()                |
      | InsertAtPos()               |
      | DeleteFirst()               |
      | DeleteLast()                |
      | DeleteAtPos()               |
      +-----------------------------+

===============================================================================
Explanation:
===============================================================================

This program demonstrates the use of the Scope Resolution Operator (::) to
define member functions outside the class.

Inside the class, only the function prototypes (declarations) are provided.
The actual implementation of each member function is written separately using
the syntax:

ReturnType ClassName::FunctionName()
{
    // Function Definition
}

This approach separates the interface of the class from its implementation,
making the program cleaner, easier to understand, and simpler to maintain.

The constructor initializes the linked list by setting first to NULL and
iCount to 0.

===============================================================================
Program Flow:
===============================================================================

Declare SinglyLL Class
          │
          ▼
Declare Member Functions
          │
          ▼
Define Functions using ::
          │
          ▼
Create SinglyLL Object
          │
          ▼
Constructor Invoked
          │
          ▼
Initialize first = NULL
Initialize iCount = 0
          │
          ▼
Program Terminates

===============================================================================
Operations Performed:
===============================================================================

1. Defined a self-referential node structure.
2. Created the SinglyLL class.
3. Declared all member functions inside the class.
4. Defined member functions outside the class.
5. Used the Scope Resolution Operator (::).
6. Initialized first and iCount in the constructor.
7. Created an object of the SinglyLL class.

===============================================================================
Advantages of Scope Resolution Operator:
===============================================================================

• Separates declaration from implementation.
• Improves code readability.
• Makes programs easier to maintain.
• Supports modular programming.
• Widely used in professional C++ development.

===============================================================================
Time Complexity:
===============================================================================

Constructor      : O(1)
Count()          : O(1)
Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 008 demonstrates how the Scope Resolution Operator (::) is used to
define member functions outside the class. This programming style improves
code organization, promotes modularity, and is considered a standard practice
for developing large-scale Object-Oriented C++ applications.

===============================================================================
*/
